lib/mix/lib/mix/tasks/deps.clean.ex

```diff
@@ -29,6 +29,14 @@ defmodule Mix.Tasks.Deps.Clean do
     Mix.Project.get!()
     {opts, apps} = OptionParser.parse!(args, strict: @switches)
 
+    Mix.Project.with_build_lock(fn ->
+      Mix.Project.with_deps_lock(fn ->
+        do_run(args, opts, apps)
+      end)
+    end)
```

Comment:
We need to be careful about deadlocks, like if any of the deps tasks were to call compile, but I think that's not the case, so currently we should be good?

Reply:
I would refactor the code and make sure that the deps manipulation and build manipulation happen in distinct places and they don't overlap!

Reply:
Sounds good, should we print `* Cleaning foo` only for the build ones, or do we print twice for every app?

Reply:
Let's print it once, for whatever we do last (probably for the build one, but any is fine).

Reply:
Build makes more sense, because deps/ can be skipped if `--build` is passed. We do build first, but I think that's fine (cleaning build first makes sense, it's reverse order for get and compile).